src/Domain/BoundaryConditions/BoundaryCondition.hpp

```diff
@@ -0,0 +1,34 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <charm++.h>
+#include <memory>
+#include <pup.h>
+#include <string>
+
+#include "Parallel/CharmPupable.hpp"
+
+/// \ingroup ComputationalDomainGroup
+/// \brief %Domain support for applying boundary conditions
+namespace domain::boundary_conditions {
+/*!
+ * \brief Base class off of which all system-specific base classes must inherit.
+ */
+class BoundaryCondition : public PUP::able {
+ public:
+  BoundaryCondition() = default;
+  BoundaryCondition(BoundaryCondition&&) noexcept = default;
+  BoundaryCondition& operator=(BoundaryCondition&&) noexcept = default;
+  BoundaryCondition(const BoundaryCondition&) = default;
+  BoundaryCondition& operator=(const BoundaryCondition&) = default;
```

Comment:
Make all constructors `protected`? I think we do that in other abstract base classes

Reply:
It's a pure virtual base class so there's no gain in marking the constructors protected. You can't construct an abstract base class anyway. The only places in the code that uses this pattern is https://github.com/sxs-collaboration/spectre/blob/eed5ea24884be93176b8bc3a2a0278183ef08afc/src/NumericalAlgorithms/LinearSolver/LinearSolver.hpp, and it's not necessary there either. The other places that protect constructors are base classes that shouldn't even be instantiated on their own but aren't abstract base classes, the place where this is actually required.

Reply:
In the linear solver code I copied this pattern from `Event`. `Event` uses the pattern because it has no pure virtual member so it hides the constructors. I thought it would be good to always do that in abstract base classes, just to make it easier for people who copy the code. Up to you.

Reply:
I don't see why we should insist on unnecessary restrictions to classes. I personally think it's actually easier for people if they aren't marked `protected` because then people won't be confused about why they are protected